File: server/src/game/effects/EffectContext.ts
import type { Game } from '../Game.js';
import type { Minion } from '../Minion.js';
import type { Player } from '../Player.js';

// Polymorphism context — mọi effect cùng giải quyết qua execute(context).
export interface EffectContext {
  game: Game;
  player: Player;
  opponent: Player;
  target?: Minion | Player | null;
  value: number;
  areaTargets?: readonly Minion[];
}

File: server-java/src/main/java/vn/coincard/server/game/effects/EffectContext.java
package vn.coincard.server.game.effects;

import java.util.List;
import vn.coincard.server.game.Minion;
import vn.coincard.server.game.Player;
import vn.coincard.server.model.GameCharacter;

/** Inputs available to one card-effect strategy. */
public class EffectContext {
  public final Player player;
  public final Player opponent;
  public final GameCharacter target; // Hero | Minion | null
  public final List<Minion> areaTargets; // for AOE_* targets, may be null

  public EffectContext(Player player, Player opponent, GameCharacter target, List<Minion> areaTargets) {
    this.player = player;
    this.opponent = opponent;
    this.target = target;
    this.areaTargets = areaTargets;
  }
}
